04/app.js

```diff
@@ -1,87 +1,96 @@
-import React from 'react';
-import ReactDOM from 'react-dom';
+import React from "react";
+import ReactDOM from "react-dom";
 
 class App extends React.Component {
-    state = { 
-        firstName: '',
-        lastName: '',
-        searchQuery: '',
-        users: ['Jan Kowalski', 'Michał Nowak'],
-    }
+  state = {
+    firstName: "",
+    lastName: "",
+    searchQuery: "",
+    users: ["Jan Kowalski", "Michał Nowak"],
+  };
 
-    renderUsersList() {
-        const {users} = this.state;
-        return users.map(name => {
-            return (
-                <li onClick={ this.clickHandler }>
-                    { name }
-                </li>
-            );
-        });
-    }
+  renderUsersList() {
+    const { users, searchQuery } = this.state;
+    return users.map((name, index) => {
+      if (searchQuery.trim().length > 0) {
+        if (!name.toLowerCase().includes(searchQuery.toLowerCase())) {
+          return null;
+        }
+      }
 
-    clickHandler = e => {
-        const {innerText: userName} = e.target;
-        this.removeUser(userName);
-    }
+      return (
+        <li key={index} onClick={this.clickHandler}>
+          {name}
+        </li>
+      );
+    });
+  }
 
-    inputChange = e => {
-        const {name, value} = e.target;
-        this.setState({
-            [name]: value,
-        });
-    }
+  clickHandler = (e) => {
+    const { innerText: userName } = e.target;
+    this.removeUser(userName);
+  };
 
-    render() {
-        const { firstName, lastName } = this.state;
-        return (
-            <section onSubmit={ this.submitHandler }>
-                <form>
-                    <input name="firstName"
-                        value={ firstName }
-                        onChange={ this.inputChange }
-                    />
-                    <input name="lastName"
-                        value={ lastName }
-                        onChange={ this.inputChange }
-                    />
-                    <input type="submit"/>
-                </form>
-                <ul>{ this.renderUsersList() }</ul>
-            </section>
-        );
-    }
+  inputChange = (e) => {
+    const { name, value } = e.target;
+    this.setState({
+      [name]: value,
+    });
+  };
 
-    submitHandler = e => {
-        e.preventDefault();
+  render() {
+    const { firstName, lastName, searchQuery } = this.state;
+    return (
+      <section onSubmit={this.submitHandler}>
+        <form>
+          <input
+            name="firstName"
+            value={firstName}
+            onChange={this.inputChange}
+          />
+          <input name="lastName" value={lastName} onChange={this.inputChange} />
+          <input type="submit" />
+        </form>
+        <input
+          type="text"
+          name="searchQuery"
+          placeholder="Search user"
+          value={searchQuery}
+          onChange={this.inputChange}
+        />
+        <ul>{this.renderUsersList()}</ul>
```

Comment:
Mógłbyś sprawdzać ile jest wyrenderowanych (widocznych) elementów, a ile jest w `state`.
To pozwoli Ci obliczyć jaka jest różnica.
Dodatkowo mógłbyś zapisywać czas dodania do state w formie timestamp co pozwoli Ci rozpoznać czy dane element jest "nowy". W tym wypadku w `this.users` będziesz miał: `[{name: 'Jan Kowalski', time: 32432432}]`, podobnie w searchQuery ;)